pymc3/diagnostics.py

```diff
@@ -263,13 +254,36 @@ def calc_n_eff(x):
 
     n_eff = {}
     for var in mtrace.varnames:
-
-        # Get all traces for var
         x = np.array(mtrace.get_values(var, combine=False))
 
-        try:
-            n_eff[var] = calc_n_eff(x)
-        except ValueError:
-            n_eff[var] = [calc_n_eff(y.transpose()) for y in x.transpose()]
+        # make sure to handle scalars correctly - add extra dim if needed
+        if len(x.shape) == 2:
+            is_scalar = True
+            x = np.atleast_3d(mtrace.get_values(var, combine=False))
+        else:
+            is_scalar = False
+
+        # now we are going to transpose all dims - makes the loop below
+        # easier by moving the axes of the variable to the front instead
+        # of the chain and sample axes
+        x = x.transpose()
+
+        Vhat = get_vhat(x)
+
+        # get an array the same shape as the var
+        _n_eff = np.zeros(x.shape[:-2])
+
+        # iterate over tuples of indices of the shape of var
+        inds = [y.ravel().tolist() for y in np.indices(x.shape[:-2])]
+        for tup in zip(*inds):  # iterate with zip
+            _n_eff[tup] = get_neff(x[tup], Vhat[tup])
```

Comment:
For variables with very large sizes, this way of building the inds is not great since they all have to be put into memory. However, I don't expect this to be so common but what do I know...